webapp/src/components/App/Studies/StudyTree/index.tsx

```diff
@@ -0,0 +1,298 @@
+        {
+          path: `${folder.parentPath}/${folder.name}`,
+          name: folder.name,
+          children: [],
+        },
+      ],
+    };
+  }
+
+  // not a direct child, but does belong to this branch so recursively walk though the tree
+  return {
+    ...studiesTree,
+    children: studiesTree.children.map((child) =>
+      insertFolderIfNotExist(child, folder),
+    ),
+  };
+}
+
+/**
+ * Insert several folders in the study tree if they don't exist already in the tree.
+ *
+ * @param studiesTree study tree to insert the folder into
+ * @param folders folders to inert into the tree
+ * @param studiesTree study tree to insert the folder into
+ * @param folder folder to inert into the tree
+ * @returns study tree with the folder inserted if it wasn't already there.
+ * New branch is created if it contain the folder otherwise the branch is left unchanged.
+ */
+export function insertFoldersIfNotExist(
+  studiesTree: StudyTreeNode,
+  folders: NonStudyFolder[],
+): StudyTreeNode {
+  return folders.reduce(
+    (tree, folder) => insertFolderIfNotExist(tree, folder),
+    studiesTree,
+  );
+}
+
+/**
+ * Call the explorer api to fetch the subfolders under the given path.
+ *
+ * @param path path of the subfolder to fetch, should sart with root, e.g. root/workspace/folder1
+ * @returns list of subfolders under the given path
+ */
+async function fetchSubfolders(path: string): Promise<NonStudyFolder[]> {
+  if (path === "root") {
+    // Under root there're workspaces not subfolders
+    return [];
+  }
+  // less than 2 parts means we're at the root level
+  const pathParts = path.split("/");
+  if (pathParts.length < 2) {
+    return [];
+  }
+  // path parts should be ["root", workspace, "folder1", ...]
+  const workspace = pathParts[1];
+  const subPath = pathParts.slice(2).join("/");
+  return await api.getFolders(workspace, subPath);
+}
+
+/**
+ * Fetch and insert the subfolders under the given paths into the study tree.
+ *
+ * This function is used to fill the study tree when the user clicks on a folder.
+ *
+ * Subfolders are inserted only if they don't exist already in the tree.
+ *
+ * @param paths list of paths to fetch the subfolders for
+ * @param studiesTree study tree to insert the subfolders into
+ * @returns study tree with the subfolders inserted if they weren't already there.
+ */
+async function fetchAndInsertSubfolders(
+  paths: string[],
+  studiesTree: StudyTreeNode,
+): Promise<[StudyTreeNode, string[]]> {
+  return paths.reduce(
+    async (acc, path) => {
+      const [accTree, accFailedPaths] = await acc;
+      try {
+        const subfolders = await fetchSubfolders(path);
+        return [insertFoldersIfNotExist(accTree, subfolders), accFailedPaths];
+      } catch (error) {
+        console.error("failed to load path ", path, error);
+        return [accTree, [...accFailedPaths, path]];
+      }
+    },
+    Promise.resolve<[StudyTreeNode, string[]]>([studiesTree, []]),
+  );
+}
```

Comment:
The current code processes paths sequentially through an async reduce, forcing each path to wait for the previous one to complete. This creates unnecessary delays that compound with more paths. Error handling is also inefficient, with try/catch blocks cascading through the promise chain and complicating state management.

To avoid this you can use parallel fetching using `Promise.allSettled` instead of sequential async reduce, using this you can:
- continue processing even if some paths fail
- track which specific paths failed
- still get results from successful paths

```typescript
async function fetchAndInsertSubfolders(
  paths: string[],
  studiesTree: StudyTreeNode,
): Promise<[StudyTreeNode, string[]]> {
  const results = await Promise.allSettled(
    paths.map((path) => fetchSubfolders(path)),
  );

  return results.reduce<[StudyTreeNode, string[]]>(
    ([tree, failed], result, index) => {
      if (result.status === "fulfilled") {
        return [insertFoldersIfNotExist(tree, result.value), failed];
      }
      console.error("Failed to load path:", paths[index], result.reason);
      return [tree, [...failed, paths[index]]];
    },
    [studiesTree, []],
  );
}
```